src/components/frontend-engine/use-form-change.ts

```diff
@@ -0,0 +1,154 @@
+	};
+
+	// =============================================================================
+	// EFFECTS
+	// =============================================================================
+	useEffect(() => {
+		const subscription = watch((value, { name }) => {
+			if (name) {
+				setField(name, value[name]);
+			} else {
+				setFields(value);
+			}
+		});
+		return () => subscription.unsubscribe();
+	}, []);
+
+	useEffect(() => {
+		// attach / fire onChange event only when formValidationConfig has values
+		// otherwise isValid will be returned incorrectly as true
+		if (onChange && Object.keys(formValidationConfig || {}).length) {
+			const subscription = watch(() => {
+				onChange(getFormValues(undefined, stripUnknown), checkIsFormValid());
+			});
+			onChange(getFormValues(undefined, stripUnknown), checkIsFormValid());
+
+			return () => subscription.unsubscribe();
+		}
+
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [checkIsFormValid, onChange, watch, formValidationConfig]);
+
+	useEffect(() => {
+		const subscription = watch(() => {
+			handleValueChange.current();
+		});
+
+		return () => subscription.unsubscribe();
+	}, [watch]);
+
+	useEffect(() => {
+		// when config changes due to overrides or conditional rendering, mark validity for re-evaluation
+		if (previousFormValidationConfig) {
+			const mapValidationRules = (config: TFormYupConfig) => {
+				return Object.entries(config).map(([fieldId, { validationRules }]) => [fieldId, validationRules]);
+			};
+			const previousValidationRules = mapValidationRules(previousFormValidationConfig);
+			const validationRules = mapValidationRules(formValidationConfig);
+			if (!isEqual(previousValidationRules, validationRules)) {
+				setHasSchemaChange(true);
+			}
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [formValidationConfig]);
+
+	useEffect(() => {
+		// re-evaluate form validity when schema changes
+		if (hasSchemaChange && previousHardValidationSchema !== hardValidationSchema) {
+			handleValueChange.current();
+			setHasSchemaChange(false);
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [hasSchemaChange, hardValidationSchema]);
+
+	useEffect(() => {
+		const errors = formState.errors;
+
+		if (errors && !isEmpty(errors)) {
+			const subscription = watch((value) => {
+				const apiErrors = Object.fromEntries(
+					Object.entries(formState.errors).filter(([_, value]) => value.type === "api")
+				);
+				const hasApiErrors = apiErrors && !isEmpty(apiErrors);
+
+				if (hasApiErrors) {
+					Object.keys(apiErrors).forEach((key) => {
+						const oldValue = oldFormValues[key];
+						const updatedValue = value[key];
+
+						if (oldValue !== updatedValue) {
+							clearErrors(key);
+						}
+					});
+				}
+
+				setOldFormValues(value);
+			});
+
+			return () => subscription.unsubscribe();
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [formState, watch]);
```

Comment:
why move this useEffect here? I feel that it has more to do with field errors than form changes

Reply:
the error handling is also in response to form changes via `watch`, so I thought it made sense to put them together. I can revert this if it's preferred

Reply:
I see, alright we keep this here then, we will move this out when there are enough error handling logic to warrant a separate file